go/runtime/connector_proxy.go

```diff
@@ -0,0 +1,196 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/estuary/flow/go/bindings"
+	pc "github.com/estuary/flow/go/protocols/capture"
+	pd "github.com/estuary/flow/go/protocols/derive"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	pm "github.com/estuary/flow/go/protocols/materialize"
+	"github.com/estuary/flow/go/protocols/ops"
+	pr "github.com/estuary/flow/go/protocols/runtime"
+	pb "go.gazette.dev/core/broker/protocol"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type connectorProxy struct {
+	address   pb.Endpoint
+	host      *FlowConsumer
+	mu        sync.Mutex
+	runtimes  map[string]*grpc.ClientConn
+	semaphore chan struct{}
+}
+
+func (s *connectorProxy) ProxyConnectors(stream pr.ConnectorProxy_ProxyConnectorsServer) error {
+	var _, cancel, _, err = s.host.Service.Verifier.Verify(stream.Context(), pf.Capability_PROXY_CONNECTOR)
+
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	// Use `semaphore` to constrain the total number of allowed proxy runtimes.
+	s.semaphore <- struct{}{}
+	defer func() { <-s.semaphore }()
+
+	// Unique id for this proxy, that we'll pass back to the client.
+	var id = fmt.Sprintf("connector-proxy-%d", time.Now().UnixNano())
+
+	_ = stream.Send(&pr.ConnectorProxyResponse{
+		Address: s.address,
+		ProxyId: id,
+	})
+	svc, err := bindings.NewTaskService(
+		pr.TaskServiceConfig{
+			AllowLocal:       s.host.Config.Flow.AllowLocal,
+			ContainerNetwork: s.host.Config.Flow.Network,
+			TaskName:         id,
+			UdsPath:          path.Join(os.TempDir(), id),
+		},
+		func(log ops.Log) {
+			_ = stream.Send(&pr.ConnectorProxyResponse{Log: &log})
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	s.runtimes[id] = svc.Conn()
+	s.mu.Unlock()
+
+	// Block until we read EOF, signaling a graceful shutdown.
+	for {
+		if _, err = stream.Recv(); err == nil {
+			// Pass.
+		} else if err == io.EOF {
+			break
+		} else {
+			return err
```

Comment:
i'm realizing the error handling here needs to ensure we're still tearing down `svc` (and we should log, too). Doing that.